diff --git a/src/dnsmasq.h b/src/dnsmasq.h
index d2823ba..65b62ca 100644
--- a/src/dnsmasq.h
+++ b/src/dnsmasq.h
@@ -251,7 +251,8 @@ struct event_desc {
 #define OPT_MAC_HEX        55
 #define OPT_TFTP_APREF_MAC 56
 #define OPT_METRIC_ORDER   57
-#define OPT_LAST           58
+#define OPT_FILTER_AAAA    58
+#define OPT_LAST           59
 
 /* extra flags for my_syslog, we use a couple of facilities since they are known 
    not to occupy the same bits as priorities, no matter how syslog.h is set up. */
diff --git a/src/option.c b/src/option.c
index c9a1925..3ddf5ce 100644
--- a/src/option.c
+++ b/src/option.c
@@ -161,6 +161,7 @@ struct myoption {
 #define LOPT_TFTP_MTU      349
 #define LOPT_REPLY_DELAY   350
 #define LOPT_METRIC_ORDER  351
+#define LOPT_FILTER_AAAA   352
  
 #ifdef HAVE_GETOPT_LONG
 static const struct option opts[] =  
@@ -324,6 +325,7 @@ static const struct myoption opts[] =
     { "quiet-dhcp6", 0, 0, LOPT_QUIET_DHCP6 },
     { "quiet-ra", 0, 0, LOPT_QUIET_RA },
     { "dns-loop-detect", 0, 0, LOPT_LOOP_DETECT },
+    { "filter-aaaa", 0, 0, LOPT_FILTER_AAAA },
     { "script-arp", 0, 0, LOPT_SCRIPT_ARP },
     { "dhcp-ttl", 1, 0 , LOPT_DHCPTTL },
     { "dhcp-reply-delay", 1, 0, LOPT_REPLY_DELAY },
@@ -500,6 +502,7 @@ static struct {
   { LOPT_IGNORE_ADDR, ARG_DUP, "<ipaddr>", gettext_noop("Ignore DNS responses containing ipaddr."), NULL }, 
   { LOPT_DHCPTTL, ARG_ONE, "<ttl>", gettext_noop("Set TTL in DNS responses with DHCP-derived addresses."), NULL }, 
   { LOPT_REPLY_DELAY, ARG_ONE, "<integer>", gettext_noop("Delay DHCP replies for at least number of seconds."), NULL },
+  { LOPT_FILTER_AAAA, OPT_FILTER_AAAA, NULL, gettext_noop("Filter all AAAA requests."), NULL },
   { 0, 0, NULL, NULL, NULL }
 }; 
 
diff --git a/src/rfc1035.c b/src/rfc1035.c
index b078b59..79dcd89 100644
--- a/src/rfc1035.c
+++ b/src/rfc1035.c
@@ -1565,6 +1565,16 @@ size_t answer_request(struct dns_header *header, char *limit, size_t qlen,
 	        break;
 #endif
 	      
+          //patch to filter aaaa forwards
+          //found : https://gist.github.com/bearice/7d3dc0e63e003d752622
+          if (qtype == T_AAAA && option_bool(OPT_FILTER_AAAA) ){
+              //return a null reply
+              ans = 1;
+              if (!dryrun) log_query(F_CONFIG | F_IPV6 | F_NEG, name, &addr, NULL);
+              break;
+          }
+          //end of patch
+          
 	      if (qtype != type && qtype != T_ANY)
 		continue;
 	      
