diff --git a/src/rfc2131.c b/src/rfc2131.c
index a679470..6033044 100644
--- a/src/rfc2131.c
+++ b/src/rfc2131.c
@@ -1200,6 +1200,9 @@ size_t dhcp_reply(struct dhcp_context *context, char *iface_name, int int_index,
 	  
 	  if (!(context = narrow_context(context, mess->yiaddr, tagif_netid)))
 	    {
+ 	      /* Do not return DHCPNAK on static network) */
+ 	      log_packet("Not sending DHCPNAK", &mess->yiaddr, emac, emac_len, iface_name, NULL, NULL, mess->xid);
+ 	      return 0;
 	      /* If a machine moves networks whilst it has a lease, we catch that here. */
 	      message = _("wrong network");
 	      /* ensure we broadcast NAK */
