diff --git a/mud/mud.c b/mud/mud.c
index 2b12d6c..d64774f 100644
--- a/mud/mud.c
+++ b/mud/mud.c
@@ -734,11 +734,11 @@ mud_keyx(struct mud *mud, unsigned char *public, int aes)
 
     if (mud->crypto.next.aes) {
         crypto_aead_aes256gcm_beforenm((crypto_aead_aes256gcm_state *)
-                                           mud->crypto.next.encrypt.state,
+                                           &mud->crypto.next.encrypt.state,
                                        mud->crypto.next.encrypt.key);
 
         crypto_aead_aes256gcm_beforenm((crypto_aead_aes256gcm_state *)
-                                           mud->crypto.next.decrypt.state,
+                                           &mud->crypto.next.decrypt.state,
                                        mud->crypto.next.decrypt.key);
     }
 }
