diff --git a/src/hostmon.c b/src/hostmon.c
index 14df2c8..6571562 100644
--- a/src/hostmon.c
+++ b/src/hostmon.c
@@ -31,7 +31,7 @@ details.
 #include <linux/if_packet.h>
 #include <linux/if_ether.h>
 #include <linux/if_fddi.h>
-#include <linux/if_tr.h>
+#include <netinet/if_tr.h>
 #include <net/if_arp.h>
 #include <stdlib.h>
 #include <time.h>
diff --git a/src/othptab.c b/src/othptab.c
index 97771d1..a8bb536 100644
--- a/src/othptab.c
+++ b/src/othptab.c
@@ -18,7 +18,7 @@ details.
 
 #include <asm/types.h>
 #include <linux/if_ether.h>
-#include <linux/if_tr.h>
+#include <netinet/if_tr.h>
 #include <linux/if_fddi.h>
 #include <winops.h>
 #include "arphdr.h"
diff --git a/src/packet.c b/src/packet.c
index 33fdf2a..1e2b81b 100644
--- a/src/packet.c
+++ b/src/packet.c
@@ -36,7 +36,7 @@ details.
 #include <linux/if_packet.h>
 #include <linux/if_ether.h>
 #include <linux/if_fddi.h>
-#include <linux/if_tr.h>
+#include <netinet/if_tr.h>
 #include <linux/isdn.h>
 #include <linux/sockios.h>
 #include <msgboxes.h>
diff --git a/src/tcptable.h b/src/tcptable.h
index 3e17793..d1380b5 100644
--- a/src/tcptable.h
+++ b/src/tcptable.h
@@ -23,7 +23,7 @@
 #include <linux/if_packet.h>
 #include <linux/if_ether.h>
 #include <linux/if_fddi.h>
-#include <linux/if_tr.h>
+#include <netinet/if_tr.h>
 #include <net/if.h>
 #include <netinet/ip.h>
 #include <netinet/udp.h>
diff --git a/src/tr.c b/src/tr.c
index 40c9e63..11f8045 100644
--- a/src/tr.c
+++ b/src/tr.c
@@ -7,7 +7,7 @@
  */
 
 #include <asm/types.h>
-#include <linux/if_tr.h>
+#include <netinet/if_tr.h>
 #include <netinet/in.h>
 
 unsigned int get_tr_ip_offset(unsigned char *pkt)
