diff --git a/luci/sqm-cbi.lua b/luci/sqm-cbi.lua
index 9b8af51..fd5dd53 100644
--- a/luci/sqm-cbi.lua
+++ b/luci/sqm-cbi.lua
@@ -13,6 +13,7 @@ You may obtain a copy of the License at
 $Id$
 ]]--
 
+local uci = luci.model.uci.cursor()
 local wa = require "luci.tools.webadmin"
 local fs = require "nixio.fs"
 local net = require "luci.model.network".init()
@@ -55,17 +56,48 @@ function e.write(self, section, value)
 	end
 	return Flag.write(self, section, value)
 end
--- TODO: inform the user what we just did...
 
-n = s:taboption("tab_basic", ListValue, "interface", translate("Interface name"))
+-- TODO: inform the user what we just did...
+n = s:taboption("tab_basic", ListValue, "_interface_all", translate("Interface name"))
 -- sm lifted from luci-app-wol, the original implementation failed to show pppoe-ge00 type interface names
 for _, iface in ipairs(ifaces) do
 	if not (iface == "lo" or iface:match("^ifb.*")) then
-		n:value(iface)
+		local multipath = uci:get("network", iface, "multipath")
+		if multipath == "on" or multipath == "backup" or multipath == "master" or multipath == "handover" then
+			n:value(iface, (uci:get("network", iface, "label") or "") .. " (" .. iface .. ")" )
+		else
+			n:value(iface)
+		end
 	end
 end
 n.rmempty = false
+n.write = function(self, section, value)
+	m:set(section, "interface", value)
+end
+n.cfgvalue = function(self, section)
+	return m:get(section, "interface", value)
+end
+n.optional = true
 
+-- TODO: inform the user what we just did...
+n1 = s:taboption("tab_basic", ListValue, "_interface_mptcp", translate("Interface name"))
+-- sm lifted from luci-app-wol, the original implementation failed to show pppoe-ge00 type interface names
+for _, iface in ipairs(ifaces) do
+	if not (iface == "lo" or iface:match("^ifb.*")) then
+		local multipath = uci:get("network", iface, "multipath")
+		if multipath == "on" or multipath == "backup" or multipath == "master" or multipath == "handover" then
+			n1:value(iface, (uci:get("network", iface, "label") or "") .. " (" .. iface .. ")" )
+		end
+	end
+end
+n1.rmempty = false
+n1.write = function(self, section, value)
+		m:set(section, "interface", value)
+end
+n1.cfgvalue = function(self, section)
+	return m:get(section, "interface", value)
+end
+n1.optional = true
 
 dl = s:taboption("tab_basic", Value, "download", translate("Download speed (kbit/s) (ingress) set to 0 to selectively disable ingress shaping:"))
 dl.datatype = "and(uinteger,min(0))"
@@ -86,23 +118,26 @@ if fs.stat(run_path) then
 		c:value( file )
 	end
 end
-c.default = "fq_codel"
+c.default = "sfq"
 c.rmempty = false
 
-
-
 local qos_desc = ""
 sc = s:taboption("tab_qdisc", ListValue, "script", translate("Queue setup script"))
 for file in fs.dir(path) do
 	if string.find(file, ".qos$") then
 		sc:value(file)
+		if file:find("otb") then
+			n1:depends("script", file)
+		else
+			n:depends("script", file)
+		end
 	end
 	if string.find(file, ".qos.help$") then
 		fh = io.open(path .. "/" .. file, "r")
 		qos_desc = qos_desc .. "<p><b>" .. file:gsub(".help$", "") .. ":</b><br />" .. fh:read("*a") .. "</p>"
 	end
 end
-sc.default = "simple.qos"
+sc.default = "otb.qos"
 sc.rmempty = false
 sc.description = qos_desc
 
@@ -182,11 +217,11 @@ ll = s:taboption("tab_linklayer", ListValue, "linklayer", translate("Which link
 ll:value("none", "none ("..translate("default")..")")
 ll:value("ethernet", "Ethernet with overhead: select for e.g. VDSL2.")
 ll:value("atm", "ATM: select for e.g. ADSL1, ADSL2, ADSL2+.")
-ll.default = "none"
+ll.default = "atm"
 
 po = s:taboption("tab_linklayer", Value, "overhead", translate("Per Packet Overhead (byte):"))
 po.datatype = "and(integer,min(-1500))"
-po.default = 0
+po.default = 44
 po.isnumber = true
 po.rmempty = true
 po:depends("linklayer", "ethernet")
diff --git a/platform/openwrt/sqm-uci b/platform/openwrt/sqm-uci
index 13a7a0b..82372e5 100644
--- a/platform/openwrt/sqm-uci
+++ b/platform/openwrt/sqm-uci
@@ -1,16 +1,11 @@
 
-config queue 'eth1'
-        option enabled '0'
-        option interface 'eth1'
-        option download '85000'
-        option upload '10000'
-        option qdisc 'fq_codel'
-        option script 'simple.qos'
-        option qdisc_advanced '0'
-        option ingress_ecn 'ECN'
-        option egress_ecn 'ECN'
-        option qdisc_really_really_advanced '0'
-        option itarget 'auto'
-        option etarget 'auto'
-        option linklayer 'none'
+config queue
+	option enabled '0'
+	option qdisc 'sfq'
+	option script 'otb.qos'
+	option linklayer 'atm'
+	option overhead '44'
+	option download '0'
+	option upload '1000'
+	option qdisc_advanced '0'

diff --git a/src/run-openwrt.sh b/src/run-openwrt.sh
index cabaa8c..3e2d0f8 100644
--- a/src/run-openwrt.sh
+++ b/src/run-openwrt.sh
@@ -76,7 +76,8 @@ run_sqm_scripts() {
     local section="$1"
     export IFACE=$(config_get "$section" interface)
 
-    [ -z "$RUN_IFACE" -o "$RUN_IFACE" = "$IFACE" ] || return
+    [ -z "$RUN_IFACE" -o "$RUN_IFACE" = "$IFACE" -o "$RUN_IFACE" = "tun0" ] || return
+    export RUN_IFACE=${RUN_IFACE}
 
     [ $(config_get "$section" enabled) -ne 1 ] && ACTION=stop
 
diff --git a/src/otb.qos b/src/otb.qos
new file mode 100644
index 0000000..2384ace
--- /dev/null
+++ b/src/otb.qos
@@ -0,0 +1,239 @@
+#!/bin/sh
+
+# This program is free software modified by Bash; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation.
+
+. ${SQM_LIB_DIR}/defaults.sh
+
+
+
+egress() {
+
+    if [ "$(uci -q get vtund.tunnel.enable)" == "1" ]; then
+        TUN0_PORT=$(uci -q get vtund.tunnel.port || echo "5005")
+    fi
+    if [ "$(uci -q get vtund.xtun0.enable)" == "1" ]; then
+	 XTUN0_PORT=$(uci -q get vtund.xtun0.port || echo "5008")
+    fi
+    if [ "$(uci -q get glorytun.otb.enable)" == "1" ]; then
+        TUN0_PORT=$(uci -q get glorytun.otb.port || echo "5006")
+    fi
+    if [ "$(uci -q get glorytun.xtun0.enable)" == "1" ]; then
+        XTUN0_PORT=$(uci -q get glorytun.xtun0.port || echo "5009")
+    fi
+    SOCKS_PORT=$(uci -q get shadowsocks.proxy.port || echo "1090")
+
+    ipt -t mangle -F QOS_MARK_${IFACE}
+    ipt -t mangle -X QOS_MARK_${IFACE}
+    ipt -t mangle -N QOS_MARK_${IFACE}
+
+    ipt -t mangle -F QOS_RULES_${IFACE}
+    ipt -t mangle -X QOS_RULES_${IFACE}
+    ipt -t mangle -N QOS_RULES_${IFACE}
+
+# Marking dscp SS tagged packets  
+    ipt -t mangle -D OUTPUT -o ${IFACE} -g QOS_MARK_${IFACE} 2>/dev/null
+    ipt -t mangle -I OUTPUT 2 -o ${IFACE} -g QOS_MARK_${IFACE}
+    ipt -t mangle -A QOS_MARK_${IFACE} -j CONNMARK --restore-mark --nfmask ${IPT_MASK} --ctmask ${IPT_MASK}
+    ipt -t mangle -A QOS_MARK_${IFACE} -m mark --mark 0x00${IPT_MASK_STRING} -j QOS_RULES_${IFACE}
+    
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs1 -m comment --comment "Scavenger" -j MARK --set-mark 0x01${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs2 -m comment --comment "Normal" -j MARK --set-mark 0x02${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs3 -m comment --comment "Signaling" -j MARK --set-mark 0x03${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs4 -m comment --comment "Realtime" -j MARK --set-mark 0x04${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs5 -m comment --comment "BroadcastVideo" -j MARK --set-mark 0x05${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs6 -m comment --comment "NetworkControl" -j MARK --set-mark 0x06${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs7 -m comment --comment "Reserved" -j MARK --set-mark 0x07${IPT_MASK_STRING}
+
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs0 -m length --length 512: -m comment --comment "Heavy" -j MARK --set-mark 0x01${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs0 -m length --length 128:512 -m comment --comment "Interactive" --j MARK --set-mark 0x03${IPT_MASK_STRING}
+    ipt -t mangle -A QOS_RULES_${IFACE} -p tcp -m multiport --dports ${SOCKS_PORT} -m dscp --dscp-class cs0 -m length --length 0:128 -m comment --comment "Acknowledge" -j MARK --set-mark 0x06${IPT_MASK_STRING}
+
+    ipt -t mangle -A QOS_MARK_${IFACE} -j CONNMARK --save-mark --nfmask ${IPT_MASK} --ctmask ${IPT_MASK}
+
+########################
+
+# Delete existing queue
+    $TC qdisc del dev ${IFACE} root 2> /dev/null
+# Traffic classes:
+# 1:2 Interactive (SSH, DNS, ACK)
+# 1:3 Low latency (VoIP)
+# 1:4 Browsing (HTTP, HTTPs)
+# 1:5 Default
+# 1:6 Low priority (p2p, pop3, smtp, etc)
+
+# add HFSC root qdisc
+    $TC qdisc add dev ${IFACE} root handle 1: `get_stab_string` hfsc default 5
+
+
+# add main rate limit class
+    $TC class add dev ${IFACE} parent 1: classid 1:1 hfsc \
+                sc rate ${UPLINK}kbit ul rate ${UPLINK}kbit
+
+# Interactive traffic: guarantee realtime full uplink for 100ms, then
+# 50% of the uplink
+
+    $TC class add dev ${IFACE} parent 1:1  classid 1:2 hfsc \
+    rt m1   ${UPLINK}kbit d  100ms m2 $((5*$UPLINK/10))kbit \
+    ls m1   ${UPLINK}kbit d  100ms m2 $((7*$UPLINK/10))kbit \
+    ul rate ${UPLINK}kbit
+
+# Interactive traffic2: guarantee realtime full uplink for 50ms, then
+# 30% of the uplink
+
+    $TC class add dev ${IFACE} parent 1:1  classid 1:3 hfsc \
+    rt m1   ${UPLINK}kbit d  50ms m2 $((3*$UPLINK/10))kbit \
+    ls m1   ${UPLINK}kbit d  50ms m2 $((5*$UPLINK/10))kbit \
+    ul rate ${UPLINK}kbit
+
+
+
+# VoIP: guarantee full uplink for 200ms, then 30%
+    $TC class add dev ${IFACE} parent 1:1  classid 1:4 hfsc \
+    sc m1 ${UPLINK}kbit d 200ms m2 $((3*$UPLINK/10))kbit \
+    ul rate ${UPLINK}kbit
+
+# Browsing: Don't guarantee anything for the first second, then
+# guarantee 10%
+
+    $TC class add dev ${IFACE} parent 1:1  classid 1:5 hfsc \
+    sc m1         0 d    1s m2 $((1*$UPLINK/10))kbit \
+    ul rate ${UPLINK}kbit
+
+# Default traffic: don't guarantee anything for the first two seconds,
+# then guarantee 5%
+
+    $TC class add dev ${IFACE} parent 1:1  classid 1:6 hfsc \
+    sc m1         0 d    2s m2 $((1*$UPLINK/20))kbit \
+    ul rate ${UPLINK}kbit
+
+# Default traffic: don't guarantee anything for the first 10 seconds,
+# then guarantee 5%
+
+    $TC class add dev ${IFACE} parent 1:1  classid 1:7 hfsc \
+    sc m1         0 d   10s m2 $((1*$UPLINK/20))kbit \
+    ul rate ${UPLINK}kbit
+
+################
+
+    $TC qdisc add dev ${IFACE} parent 1:2 handle 12: ${QDISC} \
+    `get_limit ${ELIMIT}` `get_target "${ETARGET}" ${UPLINK}` `get_ecn ${EECN}` \
+    `get_quantum  300` `get_flows ${PRIO}` ${EQDISC_OPTS}
+    $TC qdisc add dev ${IFACE} parent 1:3 handle 13: ${QDISC} \
+    `get_limit ${ELIMIT}` `get_target "${ETARGET}" ${UPLINK}` `get_ecn ${EECN}` \
+    `get_quantum  300` `get_flows ${NORMAL}` ${EQDISC_OPTS}
+    $TC qdisc add dev ${IFACE} parent 1:4 handle 14: ${QDISC} \
+    `get_limit ${ELIMIT}` `get_target "${ETARGET}" ${UPLINK}` `get_ecn ${EECN}` \
+    `get_quantum  300` `get_flows ${BULK}` ${EQDISC_OPTS}
+    $TC qdisc add dev ${IFACE} parent 1:5 handle 15: ${QDISC} \
+    `get_limit ${ELIMIT}` `get_target "${ETARGET}" ${UPLINK}` `get_ecn ${EECN}` \
+    `get_quantum  300` `get_flows ${BULK}` ${EQDISC_OPTS}
+    $TC qdisc add dev ${IFACE} parent 1:6 handle 16: ${QDISC} \
+    `get_limit ${ELIMIT}` `get_target "${ETARGET}" ${UPLINK}` `get_ecn ${EECN}` \
+    `get_quantum  300` `get_flows ${BULK}` ${EQDISC_OPTS}
+    $TC qdisc add dev ${IFACE} parent 1:7 handle 17: ${QDISC} \
+    `get_limit ${ELIMIT}` `get_target "${ETARGET}" ${UPLINK}` `get_ecn ${EECN}` \
+    `get_quantum  300` `get_flows ${BULK}` ${EQDISC_OPTS}
+
+
+# QoS for TCP/UDP ex:45aO 4=>ipv4 5=>20bytes & hexa a0=>c5
+    # cs1=20, cs2=40, cs3=60, cs4=80, cs5=a0, cs6=c0, & cs7=e0
+
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0x20 0xff  flowid 1:7
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0x40 0xff  flowid 1:6
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0x60 0xff  flowid 1:5
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0x80 0xff  flowid 1:5
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0xa0 0xff  flowid 1:4
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0xc0 0xff  flowid 1:3
+    $TC filter add dev ${IFACE} parent 1:0 protocol ip prio 1 u32 match ip tos 0xe0 0xff  flowid 1:2
+
+    # QoS for xtun0
+    $TC filter add dev ${IFACE} parent 1:0 protocol all prio 1 u32 match ip dport ${XTUN0_PORT} 0xffff  flowid 1:3
+
+}
+
+ingress() {
+      return
+}
+
+push_to_api() {
+	if mwan3 interfaces | grep tun0 | grep online; then
+		local TOKEN METRIC WAN_IP API_IP
+
+		TOKEN=$(uci -q get overthebox.me.token)
+		METRIC=$(uci -q get network.${IFACE}.metric)
+		if [ "$1" == "stop" ]; then
+			curl -s --connect-timeout 3 -X DELETE api/qos/${METRIC}?auth_token=${TOKEN}
+		elif [ "$1" == "start" ]; then
+			WAN_IP=$(curl -s --connect-timeout 3 --interface ${IFACE} ifconfig.ovh)
+			if [ -n "${WAN_IP}" ]; then
+				curl -s --connect-timeout 3 -X PUT -H "Content-Type: application/json" --data "{ \"interface\": \"${IFACE}\", \"metric\": \"${METRIC}\", \"wan_ip\": \"${WAN_IP}\", \"downlink\": \"${DOWNLINK}\", \"uplink\": \"${UPLINK}\" }" api/qos?auth_token=${TOKEN}
+			fi
+		fi
+	fi
+}
+
+sqm_start() {
+    [ -n "${IFACE}" ] || return 1
+    push_to_api start
+    [ "${RUN_IFACE}" == "tun0" ] && return 1
+    [ -z "$DEV" ] && DEV=$( get_ifb_for_if ${IFACE} )
+    do_modules
+    verify_qdisc "hfsc" || return 1
+    verify_qdisc ${QDISC} || return 1
+    sqm_logger "Starting ${SCRIPT}"
+
+    eth_setup
+    #ipt_setup
+
+    if [ "${UPLINK}" -ne 0 ];
+    then
+        egress
+        sqm_logger "egress shaping activated"
+    else
+        sqm_logger "egress shaping deactivated"
+        $TC qdisc del dev ${IFACE} root 2> /dev/null
+    fi
+    if [ "${DOWNLINK}" -ne 0 ];
+    then
+        verify_qdisc ingress "ingress" || return 1
+        ingress
+        sqm_logger "ingress shaping activated"
+    else
+        sqm_logger "ingress shaping deactivated"
+        $TC qdisc del dev ${DEV} root 2> /dev/null
+        $TC qdisc del dev ${IFACE} ingress 2> /dev/null
+    fi
+    return 0
+}
+
+sqm_stop() {
+    push_to_api stop
+    [ "${RUN_IFACE}" == "tun0" ] && return 1
+
+    $TC qdisc del dev ${IFACE} ingress 2> /dev/null
+    $TC qdisc del dev ${IFACE} root 2> /dev/null
+    [ -n "$CUR_IFB" ] && $TC qdisc del dev ${CUR_IFB} root 2> /dev/null
+    [ -n "$CUR_IFB" ] && sqm_logger "${0}: ${CUR_IFB} shaper deleted"
+
+
+    ipt -t mangle -D POSTROUTING -o ${IFACE} -g QOS_MARK_${IFACE}
+    ipt -t mangle -F QOS_MARK_${IFACE}
+    ipt -t mangle -X QOS_MARK_${IFACE}
+    ipt -t mangle -F QOS_RULES_${IFACE}
+    ipt -t mangle -X QOS_RULES_${IFACE}  [ -n "$CUR_IFB" ] && $IP link set dev ${CUR_IFB} down
+    
+    [ -n "$CUR_IFB" ] && $IP link delete ${CUR_IFB} type ifb
+    [ -n "$CUR_IFB" ] && sqm_logger "${0}: ${CUR_IFB} interface deleted"
+}
