diff --git a/src/run-openwrt.sh b/src/run-openwrt.sh
index 97171a6..3d538b9 100644
--- a/src/run-openwrt.sh
+++ b/src/run-openwrt.sh
@@ -23,9 +23,42 @@ if [ "$ACTION" = "stop" -a -z "$RUN_IFACE" ]; then
         # $SCRIPT variables saved in there.
         [ -f "$f" ] && ( . $f; IFACE=$IFACE SCRIPT=$SCRIPT SQM_DEBUG=$SQM_DEBUG SQM_DEBUG_LOG=$SQM_DEBUG_LOG OUTPUT_TARGET=$OUTPUT_TARGET ${SQM_LIB_DIR}/stop-sqm )
     done
+    # Clear DSCP rules
+    iptables-save | grep "j DSCP" | grep "dscp_" | sed "s/-A /iptables -t mangle -w -D /g" | sh
     exit 0
 fi
 
+# Generate dscp tagging preferences
+IPT="/usr/sbin/iptables -t mangle -w"
+genrateDscp() {
+        local ipset proto src_ip src_port dest_ip $ipset dest_port class
+
+        config_get proto $1 proto all
+        config_get ipset $1 ipset
+        config_get src_ip $1 src_ip 0.0.0.0/0
+        config_get src_port $1 src_port 0:65535
+        config_get dest_ip $1 dest_ip 0.0.0.0/0
+        config_get dest_port $1 dest_port 0:65535
+        config_get class $1 class
+
+        case $proto in
+                tcp)
+                        $IPT -D PREROUTING -p $proto -s $src_ip -d $dest_ip $ipset -m multiport --sports $src_port -m multiport --dports $dest_port -m comment --comment "dscp_$1" -j DSCP --set-dscp-class $class 2>/dev/null
+                        $IPT -A PREROUTING -p $proto -s $src_ip -d $dest_ip $ipset -m multiport --sports $src_port -m multiport --dports $dest_port -m comment --comment "dscp_$1" -j DSCP --set-dscp-class $class
+                        ;;
+                udp)
+                        $IPT -D PREROUTING -p $proto -s $src_ip -d $dest_ip $ipset -m multiport --sports $src_port -m multiport --dports $dest_port -m comment --comment "dscp_$1" -j DSCP --set-dscp-class $class 2>/dev/null
+                        $IPT -A PREROUTING -p $proto -s $src_ip -d $dest_ip $ipset -m multiport --sports $src_port -m multiport --dports $dest_port -m comment --comment "dscp_$1" -j DSCP --set-dscp-class $class
+                        ;;
+                *)
+                        $IPT -D PREROUTING -p $proto -s $src_ip -d $dest_ip $ipset -m comment --comment "dscp_$1" -j DSCP --set-dscp-class $class 2>/dev/null
+                        $IPT -A PREROUTING -p $proto -s $src_ip -d $dest_ip $ipset -m comment --comment "dscp_$1" -j DSCP --set-dscp-class $class
+                        ;;
+	esac
+}
+config_load dscp
+config_foreach genrateDscp classify
+
 config_load sqm
 
 run_sqm_scripts() {
