diff --git a/noinst/libcrafter/libcrafter/crafter/Fields/BitsField.h b/noinst/libcrafter/libcrafter/crafter/Fields/BitsField.h
index 1f67dd9..628f16d 100644
--- a/noinst/libcrafter/libcrafter/crafter/Fields/BitsField.h
+++ b/noinst/libcrafter/libcrafter/crafter/Fields/BitsField.h
@@ -46,9 +46,15 @@ namespace Crafter {
 		template<size_t psize, size_t bitp>
 		struct ByteBitPack {
 			static const int pad = (8>=(psize + bitp)) ? (8 - (psize + bitp)) : (8 - bitp);
+#if __BYTE_ORDER == __LITTLE_ENDIAN
 			word : pad ;
 			word fieldm:psize;
 			word : bitp;
+#else
+			word : bitp;
+			word fieldm:psize;
+			word : pad ;
+#endif
 		};
 
 		template<size_t psize>
@@ -110,13 +116,12 @@ Crafter::BitsField<size,nbit>::BitsField(const std::string& name, size_t nword)
 template<size_t size, size_t nbit>
 void Crafter::BitsField<size,nbit>::Write(byte* raw_data) const {
 	const byte over_bytes = ((nbit%8 + size)-1)/8;
-	const byte* bytes_ptr = (const byte *)&human;
 	byte* data_ptr = raw_data + offset;
 
 	if (over_bytes == 0) {
 
 		ByteBitPack<size,nbit%8>* ptr = reinterpret_cast<ByteBitPack<size,nbit%8>*> (data_ptr);
-		ptr->fieldm = bytes_ptr[0];
+		ptr->fieldm = human;
 
 	} else if (over_bytes == 1) {
 
