diff --git a/src/tracebox/tracebox.cc b/src/tracebox/tracebox.cc
index ccc76e9..b00407c 100644
--- a/src/tracebox/tracebox.cc
+++ b/src/tracebox/tracebox.cc
@@ -29,6 +29,7 @@
 #include <vector>
 
 extern "C" {
+#include <getopt.h>
 #include <pcap.h>
 #include <ifaddrs.h>
 #include <netinet/in.h>
@@ -611,7 +612,7 @@ int main(int argc, char *argv[])
 	IPLayer *ip = NULL;
 	string err;
 
-	while ((c = getopt(argc, argv, ":i:m:s:p:d:hnv6u")) != -1) {
+	while ((c = getopt(argc, argv, "i:m:s:p:d:hnv6u")) != -1) {
 		switch (c) {
 			case 'i':
 				iface = optarg;
@@ -643,13 +644,20 @@ int main(int argc, char *argv[])
 			case 'h':
 				ret = 0;
 				goto usage;
-			case ':':
-				cerr << "missing option argument" << endl;
-			default:
+			case '?':
+				if (optopt == 'c')
+					fprintf(stderr, "Option -%c requires an argument.\n", optopt);
+				else if (isprint (optopt))
+					fprintf(stderr, "Unknown option `-%c'.\n", optopt);
+				else
+					fprintf(stderr, "Unknown option character `\\x%x'.\n", optopt);
 				goto usage;
+			default:
+				goto run;
 		}
 	}
 
+run:
 	if (getuid() != 0) {
 		fprintf(stderr, "tracebox requires superuser permissions!\n");
 		return 1;
