diff --git a/src/redir.c b/src/redir.c
index d36fe3f..3b21314 100644
--- a/src/redir.c
+++ b/src/redir.c
@@ -241,7 +241,10 @@ server_recv_cb(EV_P_ ev_io *w, int revents)
             // continue to wait for recv
             return;
         } else {
-            ERROR("server recv");
+            
+            if (errno != ECONNRESET)
+                ERROR("server recv");
+
             close_and_free_remote(EV_A_ remote);
             close_and_free_server(EV_A_ server);
             return;