diff --git a/src/redir.c b/src/redir.c
index 3809411..1317402 100644
--- a/src/redir.c
+++ b/src/redir.c
@@ -1261,7 +1261,7 @@ main(int argc, char **argv)
             listen_ctx_current = (listen_ctx_t*) malloc(sizeof(listen_ctx_t));
             listen_ctx_current = memcpy(listen_ctx_current, &listen_ctx, sizeof(listen_ctx_t));
             local_port = dscp[dscp_num-1].port;
-            listen_ctx_current->tos = dscp[dscp_num-1].dscp;
+            listen_ctx_current->tos = dscp[dscp_num-1].dscp << 2;
         }
     } while (dscp_num-- > 0);
 
